pkg/util/helm/helm_test.go

```diff
@@ -0,0 +1,388 @@
+package helm
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	mockhelmclient "github.com/devstream-io/devstream/pkg/util/helm/mock"
+	"github.com/golang/mock/gomock"
+	helmclient "github.com/mittwald/go-helm-client"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/repo"
+)
+
+var (
+	NormalError   = errors.New("normal error")
+	NotFoundError = errors.New("release name not found")
+)
+
+var mockedRelease = release.Release{Name: "test"}
+var helmParam = &HelmParam{
+	Repo{
+		Name: "helm",
+		URL:  "test1",
+	},
+	Chart{
+		ReleaseName: "helm:v1.0.0",
+		Timeout:     "1m",
+		Wait:        true,
+	},
+}
+
+type DefaultMockClient struct {
+	helmclient.Client
+}
+
+func (c *DefaultMockClient) UninstallReleaseByName(name string) error {
+	return NotFoundError
+}
+
+type DefaultMockClient2 struct {
+	helmclient.Client
+}
+
+func (c *DefaultMockClient2) UninstallReleaseByName(name string) error {
+	return NormalError
+}
+
+func TestNewHelm(t *testing.T) {
+	entry := &repo.Entry{
+		Name:                  helmParam.Repo.Name,
+		URL:                   helmParam.Repo.URL,
+		Username:              "",
+		Password:              "",
+		CertFile:              "",
+		KeyFile:               "",
+		CAFile:                "",
+		InsecureSkipTLSverify: false,
+		PassCredentialsAll:    false,
+	}
+	atomic := true
+	if !helmParam.Chart.Wait {
+		atomic = false
+	}
+	tmout, err := time.ParseDuration(helmParam.Chart.Timeout)
+	if err != nil {
+		t.Log(err)
+	}
+	spec := &helmclient.ChartSpec{
+		ReleaseName:      helmParam.Chart.ReleaseName,
+		ChartName:        helmParam.Chart.ChartName,
+		Namespace:        helmParam.Chart.Namespace,
+		ValuesYaml:       helmParam.Chart.ValuesYaml,
+		Version:          helmParam.Chart.Version,
+		CreateNamespace:  false,
+		DisableHooks:     false,
+		Replace:          true,
+		Wait:             helmParam.Chart.Wait,
+		DependencyUpdate: false,
+		Timeout:          tmout,
+		GenerateName:     false,
+		NameTemplate:     "",
+		Atomic:           atomic,
+		SkipCRDs:         false,
+		UpgradeCRDs:      helmParam.Chart.UpgradeCRDs,
+		SubNotes:         false,
+		Force:            false,
+		ResetValues:      false,
+		ReuseValues:      false,
+		Recreate:         false,
+		MaxHistory:       0,
+		CleanupOnFail:    false,
+		DryRun:           false,
+	}
+
+	// default
+	got, err := NewHelm(helmParam)
+	if err != nil {
+		t.Errorf("error: %v\n", err)
+	}
+
+	want := &Helm{
+		Entry:     entry,
+		ChartSpec: spec,
+		Client:    got.Client,
```

Comment:
Use `got.Client` to construct `want` may not make sense. This probably leave out the step of `AddOrUpdateChartRepo()`.

Reply:
fixed